Add support for  koi's secondary LCD

dodoradio opened the pull request:
This slightly rearchitects the tool to add support for several watches. Most of the catfish code is completely untouched, but significant crude changes were made to the command line parser. The command line parsing could probably be done more correctly, but I need to get a better understanding of cli11 first. 

The main issue is that this completely messes up the systemd files. I have no idea how to set the systemd files correctly so that the correct commands for the watch are executed. Maybe the machine should be set at compile time instead of as a command line argument? I'd really appreciate some help with this one. 

The final disclaimer is that this is for an unstable device that hasn't been merged yet.

dodoradio committed:
Rearchitect to add support for koi
- rearchitect for several watches

dodoradio committed:
Add LCD colour parameter for koi

dodoradio committed:
Add flag for preparing shutdown into timepiece mode
the user will then need to manually shut down the watch

LecrisUT commented:
Another way to do it is to override the same function, and have the appropriate source be defined as a cmake option
```cmake
set(Device "dummy" CACHE STRING "Device to build for")
if(Device == "catfish")
  target_source(lcd-tools catfish-tools.cpp)
elseif(...)
endif()
```
and have the upstream bb file pass the correct flag `-DDevice=${machine}`

dodoradio commented:
Ok, that's a good idea, compile-time option is probably the best. I'll figure out a way to do that and a bbappend for catfish.

FlorentRevest commented on src/koi-tools.cpp:
```diff
@@ -0,0 +1,50 @@
+	outputBuff[5] = (uint8_t) (timeNow.minute() & 0xFF);
+	outputBuff[6] = (uint8_t) ((timeNow.second() + 1) & 0xFF); //the original casio code seems to add one second here to compensate for an update delay in the subcpu
+	return Write(outputBuff);
+
```
nit: extra empty line here

FlorentRevest commented on src/koi-tools.cpp:
```diff
@@ -0,0 +1,50 @@
+}
+
+int AsteroidOS::LCD_Tools::Write(char* data) {
+	QFile mMSensorFile("/dev/MultiSensors_CD_01");
```
Indentation isn't consistent here

FlorentRevest commented on src/koi-tools.cpp:
```diff
@@ -0,0 +1,50 @@
+}
+
+int AsteroidOS::LCD_Tools::Write(char* data) {
+	QFile mMSensorFile("/dev/MultiSensors_CD_01");
```
the 'm' prefix is a convention that means "member", it is meant to denote class members. In this case you have a stack variable so multiSensorFile sounds like a more appropriate name

dodoradio replied:
the mMsensors file isn't appropriate, that's an accidental holdover from the narwhal qml module which I used as an example. 
I hope there's nothing wrong with mixing upper CamelCase (which lecris seems to have used initially) and lower camelCase (which seems to be more common in Qt context) here

FlorentRevest commented:
I agree with @LecrisUT on the build-time approach here.

dodoradio committed:
Read machine from config file
Removes the command line switch and reads the machine codename from
machine.conf config file

dodoradio commented:
after some conversation with @MagneFire  and beroset on matrix, I've implemented an example where the machine is read from machine.conf, which means the same binary is usable across both devices. I haven't implemented a common object yet.

dodoradio commented:
after pushing I just realised that it's worth a) making the help page show up when no flag/option is passed b) making the help menu items dependent on the machine. I'm going to fix that tomorrow

LecrisUT commented:
That's a good workaround. Thankfully we do not call any specific libraries for each watches so it will work just fine. Hopefully this weekend I'll have my setup ready and show you the other way I was referring to in `hw-tools` repo

MagneFire commented:
@dodoradio I've done some experimenting for `catfish` which ended up in a bit of a rewrite of the codebase as well: https://github.com/MagneFire/lcd-tools/commit/2d5f8ad121d6893808449bd0a22bbbfae75993c7 it uses your method of detecting the watch codename.

In this case a base class is implemented, which I realize is not to dissimilar to what @LecrisUT did in `hw_controls`: https://github.com/LecrisUT/hw_controls/blob/main/src/catfish/Device.hpp#L11

beroset committed:
Alter function signatures for callback

This alters the function signatures to add the codename of each watch as
another level of namespace and makes the functions void(int) to be
compatible directly with CLI11's callbacks.

Change function signature for use as callback

This matches the change to the header file in changing the function to a
void(int).

Restructure using SpiMsg type

This restructures the code to use a SpiMsg type (alias of
std::array<uint8_t, 7>) and modifies to simplify use of functions as
callbacks.  Functions not part of the interface are static.

Use more features of CLI11

This relies more heavily on the callback feature of CLI11 and simplifies
and clarifies the interface.

Restore missing return

If the file does not open, the function should return after reporting
this fact rather than trying to write to the file.

Signed-off-by: Ed Beroset <beroset@ieee.org>

dodoradio committed:
Fix timesetting on sundays
The subcpu expects weeks to start on sunday. sunday is encoded as zero.
writing 7 as the weeks parameter would cause timesetting to fail on
sundays completely, since the subcpu totally rejects a partially wrong
date format

MagneFire commented on src/catfish-tools.cpp:
```diff
@@ -1,16 +1,17 @@
 #include <MGConfItem>
 
-int AsteroidOS::LCD_Tools::SyncTime() {
+namespace AsteroidOS::LCD_Tools::Catfish {
```
It might be nicer to use: `using namespace AsteroidOS::LCD_Tools::Catfish;` here since only those functions will be used.
Similarly with the `koi` namespace.

dodoradio replied:
@beroset what do you think?

beroset replied:
If you look at the `koi-tools.cpp` file, closely, you'll see that not all of the functions in that file are in the namespace.  Specifically, the two `static` functions are deliberately excluded from the `AsteroidOS::LCI_Tools::Koi` namespace because they are not part of the interface.  For consistency, I made both header files the same in that regard.  Also, if we use templates later, the `using namespace` form can lead to hard-to-diagnose problems.  See https://stackoverflow.com/a/37630222/3191481

MagneFire commented on src/catfish-tools.h:
```diff
@@ -0,0 +1,7 @@
+#define ASTEROIDOS_CATFISH_TOOLS_H
+
+namespace AsteroidOS::LCD_Tools::Catfish {
+	void SyncTime(int);
```
None of the `SyncTime` methods actually use the `int` type. What's the use case for this? Maybe remove it?

MagneFire replied:
Nvm, this has something to do with the CLI11 callback signatures.

beroset reviewed (approved):
I think we have everything resolved.  Looks good!

FlorentRevest reviewed (approved):
I like it, thank you! I'm not merging it because it's marked as draft but I think it's fine to merge.

MagneFire commented on src/lcd-tools.cpp:
```diff
@@ -1,20 +1,31 @@
+	if (machineCodename == "koi") {
+		using namespace AsteroidOS::LCD_Tools::Koi;
+		app.add_flag("--sync-time",SyncTime,"Sync lcd time with linux time");
+		app.add_flag("--white-background",[](int){SetDisplayColor(true);},"(Koi) set display background to white");
```
I'm wondering what the added benefit is of `(Koi)` in these cases.
They might as well be removed right?

dodoradio replied:
I just wanted to highlight the watch-specific features to users. These features are implemented here for testing and as a stopgap (ideally the lcd colour switching should also be presented somewhere in UI) and I want to make it clear to developers which features will only show up on their watch so they don't make any assumptions about other devices.

MagneFire replied:
Using that logic, the LCD timesync is also unique to `koi` :wink:
The `timepiece` mode is also available in an eventual update for `catfish`.
And obviously these won't be available on any non-second display watches.

dodoradio replied:
this is LCD-tools, it's only available on watches with additional timekeeping features that need time to be synchronised, so I think it's a reasonable assumption that all devices running the tool will need the `--sync-time` function. Catfish still doesn't have timepiece mode as of this PR, so the `(koi)` label should be removed when that's implemented.

dodoradio replied:
Another issue is that narwhal (which I intend to move here from hands-timesync for consistency) doesn't have a dedicated timepiece mode, it will just keep ticking after the system is shut down, assuming the hands haven't been locked into a certain location. If narwhal is added, the `timepiece` feature will need watch-specific labeling again. This gets even more complicated if the UI _does_ move the hands around - but we'll burn that bridge when we get to it.

MagneFire replied:
Sounds good :+1:

MagneFire reviewed (commented):
Minor nitpick, otherwise looks fine to me :smile:
